File: term2/1_maxHeap.java
import java.util.*;
import java.lang.*;
class heap{

	Scanner sc = new Scanner(System.in);
	int[] heapp = new int[11];
	static int capacity = 10;
	static int N = 0;
	static int swap(int a, int b){
		return a;
	}

	static void shiftup(heap obj, int k){
		while(k > 1 && obj.heapp[parent(k)] < obj.heapp[k]){
			obj.heapp[parent(k)]=swap(obj.heapp[k], obj.heapp[k]=obj.heapp[parent(k)]);
			k = parent(k);
		}
	}

	static void shiftdown(heap obj, int k){
		while(2 * k <= N){
			int j = 2 * k;
			if(j < N && obj.heapp[j] < obj.heapp[j+1])
				j++;
			if(obj.heapp[k] < obj.heapp[j])
				obj.heapp[k]=swap(obj.heapp[j], obj.heapp[j]=obj.heapp[k]);
			k = j;
		}

	}

	static int parent(int k){
		return k/2;
	}
	
	static void changeKey(heap obj, int index, int value){         //change the specified key at a given index and display after heapifying it
		int old = obj.heapp[index];
		obj.heapp[index] = value;
		if(obj.heapp[index] > old){
			shiftup(obj, index);
		}else{
			shiftdown(obj, index);
		}
	}
	                                                             
	static void insert(heap obj, int value) {       // insert an element into the heap
		if(N == capacity){
			System.out.println(" NO SPACE LEFT");
			return;
		}
		obj.heapp[++N] = value;
		shiftup(obj, N);
		
	}

	static void increaseKey(heap obj, int index, int value) {  
		obj.heapp[index] = value;
		shiftup(obj, index);
	}


	static void extractMax(heap obj) {    // remove and display max element in heap
		if(N < 1){
			System.out.println("NO ELEMENT PRESENT");
			return;
		}
		int max = obj.heapp[1];
		obj.heapp[N--] = swap(obj.heapp[1], obj.heapp[N--]=obj.heapp[1]);
		shiftdown(obj, 1);
		System.out.println(max);
	}

	static void getMax(heap obj) {          // display max element
		System.out.println(obj.heapp[1]);
	}

	static void display(heap obj) {        // display the array 
		for(int i = 1 ; i <= N ; i++) {
			System.out.print(obj.heapp[i] + " ");
		}
		System.out.println();
	}

	public static void main(String[] args) {
		heap h = new heap();
		int opt, num, value, index;
		while(true){
			System.out.println(" Enter 1 to INSERT \n Enter 2 to EXTRACT MAX \n Enter 3 to GET MAX \n Enter 4 to DISPLAY \n Enter 5 to CHANGE KEY \n Enter 6 to INCREASE KEY \n Enter 7 to EXIT");
			opt = h.sc.nextInt();
			switch(opt){
				case 1: num = h.sc.nextInt();
				insert(h, num); 
				break;
				case 2: extractMax(h); break;
				case 3: getMax(h); break;
				case 4: display(h); break;
				case 5: value = h.sc.nextInt();
				index = h.sc.nextInt();
				changeKey(h, value, index); 
				break;
				case 6: value = h.sc.nextInt();
				index = h.sc.nextInt();
				increaseKey(h, value, index); 
				break;
				case 7: System.exit(0);
			}
		}
	}
}

File: term2/1_maxheap.cpp
#include <iostream>
using namespace std;

int N = 0, heap[11], capacity = 10;

int parent(int k){
	return k/2;
}


void shiftup(int k){
	while(k > 1 && heap[parent(k)] < heap[k]){
		swap(heap[parent(k)], heap[k]);
		k = parent(k);
	}
}

void shiftdown(int k){
	while(2 * k <= N){
		int j = 2 * k;
		if(j < N && heap[j] < heap[j+1])
			j++;
		if (heap[k] > heap[j]){
			swap(heap[k], heap[j]);
		}
		k = j;
	}
}

void display(){
	for (int i = 1; i <= N; i++) {
		cout<<heap[i]<<" ";
	}
	cout<<endl;
}

int getMax(){
	return heap[1];
}

int extractMax(){
	if(N < 1){
		cout<<"TRYING TO EXTRACT FROM EMPTY HEAP"<<endl;
		return -1;
	}
		int MAX = heap[1];
		swap(heap[1], heap[N--]);
		shiftdown(1);
		return MAX;
	
}

void ins(int num){
	if(N == capacity){
		cout<<"MAX CAPACITY REACHED"<<endl;
		return;
	}
	heap[++N] = num;
	shiftup(N);
}

void increaseKey(int index, int value){
	heap[index] = value;
	shiftup(index);
}

void changeKey(int index, int value){
	int old = heap[index];
	heap[index] = value;
	if(heap[index] > old){
		shiftup(index);
	}else if(heap[index] < old){
		shiftdown(index);
	}
}

int main(){
	int ch, num, value, index;
	while(1){
		cout<<"1.INSERT"<<endl;
		cout<<"2.GET MAX"<<endl;
		cout<<"3.DISPLAY"<<endl;
		cout<<"4.EXTRACT MAX"<<endl;
		cout<<"5.INCREASE KEY"<<endl;
		cout<<"6.CHANGE KEY"<<endl;
		cout<<"7.EXIT"<<endl;
		cout<<"ENTER CHOICE"<<endl;
		cin>>ch;
		switch(ch){
			case 1 : cout<< "ENTER NUMBER" << endl;
			cin>>num;
			ins(num);
			break;
			case 2 : cout << "MAX ELEMENT: " << getMax() << endl;
			break;
			case 3 : display();
			break;
			case 4 : cout<< "MAX ELEMENT: " << extractMax() << endl;
			break;
			case 5 : cout<< "ENTER VALUE AND INDEX" << endl;
			cin >> value >> index;
			increaseKey(index, value);
			break;
			case 6 : cout<< "ENTER VALUE AND INDEX" << endl;
			cin >> value >> index;
			changeKey(index, value);
			break;
			case 7 : exit(0);
		}
	}
}
